### ISSUE TITLE
Individual replacement 

### ISSUE THREAD

**Issue by Bruneux** [dnGrep/dnGrep#301]
It seems that, once the search is done, we cannot do individual replacement rather than replace all the terms found.

**Comment by doug24** [comment]
After searching, you select files from the results, right click, and choose 'Exclude from results' from the context menu.  Then you can replace on the remaining files.

**Comment by PaulSoderlind** [comment]
Still, it would be very nice to have a "prompt for every change/every file" feature.

**Comment by Bruneux** [comment]
But, if I understand, we cannot exclude some occurrences in a particular file, can we?

**Comment by doug24** [comment]
Ah no, there is no way to replace only some occurrences in a file.

OK, I get that people would like a feature as @PaulSoderlind described as "prompt for every change/every file".  

I'm thinking that it would be a pop-up dialog that shows the the next occurrence, and allows the user to accept or skip or replace each instance.  It could also have a button to jump to the next file, or stop replacing (skip the rest of the occurrences).

I'm wondering if it will be sufficient to show the instance with the context as it now appears in the dnGrep results view (line number and a few lines above and below), or do you need to see the whole file?  Either way, there are edge cases for very long lines and for very large files that are difficult to display.

I am concerned this would be a lot of work both for the UI, and integrating it into the way that dnGrep modifies files for the replace function.

Any comments?

**Comment by JVimes** [comment]
It's a good idea, but I bet it's lower priority than some other issues. Doug's right, it's not trivial to implement (we could gracefully degrade the experience for edge cases like long lines).

**Comment by doug24** [comment]
I've been working on this as time permits, and have been making progress:  I've had to make some pretty extensive changes to the core search and replace methods, and I have the XPath replace working again (#229).  I still have to finalize the UI for filtering search results for the replace operations, so it's not ready yet.

I've added more unit tests, but with all these changes, I'm still concerned about destabilizing the application, and I don't want to put out a release that's going to generate a bunch of new defects.  

I'd really like to add more unit tests, and that's were everyone can help:  I'm looking for contributions of search and replace tests.  I would need is a test file (or the text for a file), a search string, replace string search type, and any search options (multiline, case sensitive, etc.), and the expected results.   Anything beyond a plain text search for 'X' and replace with 'Y'.   It would be great to get variety of regular expressions, capture groups, XPath, and with different file encoding and different languages.  

Please post something here, or email me directly.  Don't include any proprietary information as it will go into the test repository.  Files under 50-100K would be ideal.

**Comment by doug24** [comment]
Fixed in 2.9.127.0
